react-context/client/app.jsx

```diff
@@ -0,0 +1,74 @@
+import React from 'react';
+import AppContext from './lib/app-context';
+import parseRoute from './lib/parse-route';
+import decodeToken from './lib/decode-token';
+import Auth from './pages/auth';
+import Home from './pages/home';
+import NotFound from './pages/not-found';
+import Navbar from './components/navbar';
+import PageContainer from './components/page-container';
+
+export default class App extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      user: null,
+      isAuthorizing: true,
+      route: parseRoute(window.location.hash)
+    };
+    this.handleSignIn = this.handleSignIn.bind(this);
+    this.handleSignOut = this.handleSignOut.bind(this);
+  }
+
+  componentDidMount() {
+    window.addEventListener('hashchange', () => {
+      this.setState({
+        route: parseRoute(window.location.hash)
+      });
+    });
+    const token = window.localStorage.getItem('react-context-jwt');
+    const user = token ? decodeToken(token) : null;
+    this.setState({ user, isAuthorizing: false });
+  }
+
+  handleSignIn(result) {
+    const { user, token } = result;
+    window.localStorage.setItem('react-context-jwt', token);
+    this.setState({ user });
+  }
+
+  handleSignOut() {
+    window.localStorage.removeItem('react-context-jwt');
+    this.setState({ user: null });
+  }
+
+  renderPage() {
+    const { path } = this.state.route;
+    if (path === '') {
+      return <Home />;
+    }
+    if (path === 'sign-in' || path === 'sign-up') {
+      return <Auth />;
+    }
+    return <NotFound />;
+  }
+
+  render() {
+    if (this.state.isAuthorizing) return null;
+    const { user, route } = this.state;
+    const { handleSignIn, handleSignOut } = this;
+    const contextValue = { user, route, handleSignIn, handleSignOut };
+    return (
+      <AppContext.Provider value={contextValue}>
+        <>
+          <Navbar />
+          <PageContainer>
+            { this.renderPage() }
+          </PageContainer>
+        </>
+      </AppContext.Provider>
+    );
+  }
+}
+
+App.contextType = AppContext;
```

Comment:
You can leave line 74 out, since App is providing context to all of its descendants.